staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go

```diff
@@ -232,3 +222,62 @@ func RecordStorageListMetrics(resource string, numFetched, numEvald, numReturned
 	listStorageNumSelectorEvals.WithLabelValues(resource).Add(float64(numEvald))
 	listStorageNumReturned.WithLabelValues(resource).Add(float64(numReturned))
 }
+
+// SetStorageMonitorGetter sets monitor getter to allow monitoring etcd stats.
+func SetStorageMonitorGetter(getter func() ([]Monitor, error)) {
+	storageSize.monitorGetter = getter
+}
+
+type Monitor interface {
+	Monitor(ctx context.Context) (StorageMetrics, error)
+	Close() error
+}
+
+type storageSizeCollector struct {
+	compbasemetrics.BaseStableCollector
+
+	monitorGetter func() ([]Monitor, error)
+}
+
+// DescribeWithStability implements compbasemetrics.StableColletor
+func (c *storageSizeCollector) DescribeWithStability(ch chan<- *compbasemetrics.Desc) {
+	ch <- compbasemetrics.NewDesc(storageSizeOpts.Name, storageSizeOpts.Help, []string{"instance"}, nil, storageSizeOpts.StabilityLevel, storageSizeOpts.DeprecatedVersion)
+}
+
+// CollectWithStability implements compbasemetrics.StableColletor
+func (c *storageSizeCollector) CollectWithStability(ch chan<- compbasemetrics.Metric) {
+	monitors, err := c.monitorGetter()
+	if err != nil {
+		return
+	}
+
+	for i, m := range monitors {
+		instance := fmt.Sprintf("etcd-%d", i)
+
+		opts := storageSizeOpts
+		opts.ConstLabels = map[string]string{"instance": instance}
+		klog.InfoS("Start collecting storage metrics", "instance", instance)
+		ch <- compbasemetrics.NewGaugeFunc(&opts, func() float64 {
```

Comment:
Not sure if this is best way to define a calculated metric, but here it is.

Reply:
NewGaugeFunc isn't really meant to be used in a custom collector since a Gauge is a collector in itself.

For instance, you could have your GaugeFunc directly registered inside of the registry with the other metrics and the function would only be called at collection time. e.g. https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#example-GaugeFunc-Simple

It might be good enough for your case, but it is less malleable than a custom collector.

For a custom collector you generally want the following:
```go
var (
  storageSizeDesc = compbasemetrics.NewDesc("apiserver_storage_size_bytes", "Size of the storage database file physically allocated in bytes.", []string{"instance"}, nil, compbasemetrics.ALPHA, "")
)

func (c *monitorCollector) Describe(ch chan<- *Desc) {
  ch <- storageSizeDesc
}

func (c *myCollector) Collect(ch chan<- *Metric) {
  size, instance := compute()
  ch <- compbasemetrics.MustNewConstMetric(myMetricDesc, compbasemetrics.GaugeValue, size, instance)
}
```

Reply:
Done